core/chains/evm/gas/models.go

```diff
@@ -337,6 +343,45 @@ func (e *evmFeeEstimator) BumpFee(ctx context.Context, originalFee EvmFee, feeLi
 	return
 }
 
+func (e *evmFeeEstimator) estimateFeeLimit(ctx context.Context, feeLimit uint64, calldata []byte, toAddress *common.Address) (estimatedFeeLimit uint64, err error) {
+	// Use provided fee limit by default if EstimateGasLimit is disabled
+	if !e.geCfg.EstimateGasLimit() {
+		return commonfee.ApplyMultiplier(feeLimit, e.geCfg.LimitMultiplier())
+	}
+	// Create call msg for gas limit estimation
+	// Skip setting Gas to avoid capping the results of the estimation
+	callMsg := ethereum.CallMsg{
+		To:   toAddress,
+		Data: calldata,
+	}
+	estimatedGas, estimateErr := e.ethClient.EstimateGas(ctx, callMsg)
+	if estimateErr != nil {
+		// Do not return error if estimate gas failed, we can still use the provided limit instead since it is an upper limit
+		e.lggr.Errorw("failed to estimate gas limit. falling back to provided gas limit.", "callMsg", callMsg, "providedGasLimit", feeLimit, "error", estimateErr)
+		estimatedFeeLimit, err = commonfee.ApplyMultiplier(feeLimit, e.geCfg.LimitMultiplier())
+		return
+	}
+	e.lggr.Debugw("estimated gas", "callMsg", callMsg, "estimatedGas", estimatedGas, "providedGasLimit", feeLimit)
+	// Return error if estimated gas without the buffer exceeds the provided gas limit
+	// Transaction is destined to run out of gas and fail
+	if estimatedGas > feeLimit {
+		e.lggr.Errorw("estimated gas exceeds provided limit", "estimatedGas", estimatedGas, "providedGasLimit", feeLimit)
+		return estimatedFeeLimit, commonfee.ErrFeeLimitTooLow
+	}
+	// Apply LimitMultiplier to the estimated gas limit
+	estimatedFeeLimit, err = commonfee.ApplyMultiplier(estimatedGas, e.geCfg.LimitMultiplier())
+	if err != nil {
+		return
+	}
+	// Fallback to the provided gas limit if the buffer causes the estimated gas limit to exceed it
+	// The provided gas limit should be used as an upper bound to avoid unexpected behavior for products
+	if estimatedFeeLimit > feeLimit {
+		e.lggr.Debugw("estimated gas limit with buffer exceeds provided limit. falling back to the provided gas limit", "estimatedGasLimit", estimatedFeeLimit, "providedGasLimit", feeLimit)
+		estimatedFeeLimit = feeLimit
+	}
+	return
```

Comment:
Should this be ` commonfee.ApplyMultiplier(feeLimit, e.geCfg.LimitMultiplier())` like line 361 ?

```suggestion
	if estimatedFeeLimit > feeLimit {
		e.lggr.Debugw("estimated gas limit with buffer exceeds provided limit. falling back to the provided gas limit", "estimatedGasLimit", estimatedFeeLimit, "providedGasLimit", feeLimit)
		estimatedFeeLimit, err = commonfee.ApplyMultiplier(feeLimit, e.geCfg.LimitMultiplier())
		return
	}
	return
```

Reply:
Good point! I should be consistent but thinking through if I should remove the earlier line where I apply this multiplier. While `EstimateGas` is enabled, the expectation now is that the limit is capped at the provided gas limit. Returning `provided gas limit * LimitMultiplier`  while `EstimateGas` is enabled might conflict with that. I'll still keep `provided gas limit * LimitMultiplier` while `EstimateGas` is disabled though to maintain the existing behavior

Reply:
Would it be a good idea to provide the flexibility to the user side, if they prefer to be capped at exact gasLimist, they can set LimitMultiplier to be 1. For chains with gasLimit that can potentially under calculates, can have > 1 multiplier to have something in between estimated and the potential low gas provided.

If this make sense, it would make more sense to have `estimatedFeeLimit` comparing with `feeLimit * multiplier`

Reply:
I had considered that too since the existing behavior transmits with `feeLimit * multiplier` which we could consider as the "provided gas limit". But then we're applying `LimitMultiplier` in two different ways when `EstimateGas` is enabled. We'd use it to raise the cap but also apply it to the estimated gas. That could lead to some weird behavior since `LimitMultiplier` would be needed as a buffer for estimated gas but it also inadvertently raises the cap which a product might not want.

Reply:
But our philosophy is that products should be sending the correct gas limit for their transactions. If they send too low of a gas limit, we'll error out to avoid wasting funds but expect them to adjust their configs correctly for a chain to avoid that from happening.